[FLINK-26336] Call cancel on deletion & clean up configmaps as well

gyfora opened the pull request:
Seperating the original commit from https://github.com/apache/flink-kubernetes-operator/pull/26 to not interfere with the refactor discussion.

Aitozi commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkUtils.java:
```diff
@@ -108,21 +111,38 @@ private static void mergeInto(JsonNode toNode, JsonNode fromNode) {
                 .cascading(true)
                 .delete();
+
+        if (deleteHaConfigmaps) {
```
This seems only handle the situation for HA based on ConfigMap. Will this also work for the HA based on ZK ?

gyfora replied:
You are absolutely correct, this only works for kubernetes HA (using configmaps). 

It's a good idea to add comments to the code and highlight in the README.

In the future we could support other HA providers but that is out of scope in this ticket.

Aitozi replied:
Make sense to me, I'm OK to merge the current shape with some more comments for clarification.  

But I have sense that we may have hard work to do the nice clean up work for other HA providers, It's a bit of out of scope of the operator responsibility or ability, maybe we should extend at the Flink to support `deleteAndCleanUpHA` ?. Do you have some inputs for this cc @wangyang0918

wangyang0918 replied:
The most appropriate way to the HA data clean up is `HighAvailabilityServices#closeAndCleanupAllData()`. It should work both for ZooKeeper and K8s HA. But I agree we could comment the limitation here and do the improvement later.

Aitozi commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/JobDeploymentStatus.java:
```diff
@@ -0,0 +1,58 @@
+                        .rescheduleAfter(REFRESH_SECONDS, TimeUnit.SECONDS);
+            case DEPLOYED_NOT_READY:
+                return UpdateControl.updateStatus(flinkDeployment)
+                        .rescheduleAfter(PORT_READY_DELAY_SECONDS, TimeUnit.SECONDS);
```
Maybe make this configurable , I will create a ticket for this.

gyfora replied:
Sounds good, I was trying to preserve the current behaviour :)

Aitozi replied:
I have create the [ticket](https://issues.apache.org/jira/browse/FLINK-26399) for this

Aitozi commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/BaseReconciler.java:
```diff
@@ -75,29 +83,45 @@ public boolean removeDeployment(FlinkDeployment flinkApp) {
+        return JobDeploymentStatus.READY;
+    }
+
+    public DeleteControl shutdownAndDelete(
```
Can we document here to clarify that what is to be deleted ? Like HA data and cluster resource ?

gyfora commented:
I discovered that we need to wait for cluster shutdown before deleting the configmaps otherwise they might be recreated. Adding this now

gyfora commented:
I have addressed the comments @wangyang0918 @tweise , let me know if there is anything else.

I had to improve the waitForClusterShutdown logic to be able to create a safe point where we can delete the configmaps without the jobmanager recreating them.

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/JobDeploymentStatus.java:
```diff
@@ -0,0 +1,58 @@
+import static org.apache.flink.kubernetes.operator.reconciler.BaseReconciler.REFRESH_SECONDS;
+
+/** Status of the Flink job deployment. */
+public enum JobDeploymentStatus {
```
nit: if this applies to session mode, then the `Job` part would be misleading. But this can be dealt with later as we will need to iterate on the states more. It's already great to have a starting point!

gyfora replied:
I could rename this to cluster deployment status, then it wouldn't be confusing in case of session

gyfora replied:
or better yet JobManager as you initially named :)

tweise commented:
@gyfora I'm going to run through some of the scenarios manually to verify. Hopefully I can add the missing tests in the next days.

wangyang0918 commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkUtils.java:
```diff
@@ -108,21 +116,109 @@ private static void mergeInto(JsonNode toNode, JsonNode fromNode) {
+                        kubernetesClient
+                                .pods()
+                                .inNamespace(namespace)
+                                .withLabel(
+                                        Constants.LABEL_TYPE_KEY, Constants.LABEL_TYPE_NATIVE_TYPE)
+                                .withLabel(
+                                        Constants.LABEL_COMPONENT_KEY,
+                                        Constants.LABEL_COMPONENT_JOB_MANAGER)
+                                .withLabel(Constants.LABEL_APP_KEY, clusterId)
```
Could be simplified by `KubernetesUtils.getJobManagerSelectors(clusterId)`

wangyang0918 reviewed (approved):
+1 after the minor comments are addressed.

tweise commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkUtils.java:
```diff
@@ -108,21 +116,109 @@ private static void mergeInto(JsonNode toNode, JsonNode fromNode) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
```
It would be good to log that shutdown was completed, otherwise the last thing we see is:
```2022-02-28 21:12:45,381 o.a.f.k.o.u.FlinkUtils         [INFO ] [default.basic-example] Waiting for cluster shutdown... (5)
2022-02-28 21:12:46,395 o.a.f.k.o.u.FlinkUtils         [INFO ] [default.basic-example] Waiting for cluster shutdown... (6)
```

tweise reviewed (approved):
Verified reconciliation scenarios.

gyfora commented on flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/utils/FlinkUtils.java:
```diff
@@ -108,21 +116,109 @@ private static void mergeInto(JsonNode toNode, JsonNode fromNode) {
+        }
+    }
+
+    /** We need this due to the buggy flink kube cluster client behaviour for now. */
```
what I meant originally by this, is that the deployment client wont let you submit a new flink job as long as there is still a service around even if marked for deletion. Maybe this is not even a bug, but in any case now the comment is irrelevant with the canged behaviour, will remove it.
